docs/design/extending-api.md

```diff
@@ -0,0 +1,227 @@
+
+## Expectations about third party objects
+
+Every object that is added to a third-party Kubernetes object store is expected to contain Kubernetes
+compatible [object metadata](../devel/api-conventions.md#metadata).  This requirement enables the
+Kubernetes API server to provide the following features:
+   * Filtering lists of objects via LabelQueries
+   * `resourceVersion`-based optimistic concurrency via compare-and-swap
+   * Versioned storage
+   * Event recording
+   * Integration with basic `kubectl` command line tooling.
+   * Watch for resource changes.
+
+The `Kind` for an instance of a third-party object (e.g. CronTab) below is expected to be
+programnatically convertible to the name of the resource using
+the following conversion.  Kinds are expected to be of the form:
+
+```
+<domain-name>/<CamelCaseKind>
+```
+
+`domain-name` is expected to be a fully qualified domain name.
+
+'CamelCaseKind' is the specific type name.
+
+To convert this into the name for the `ThirdPartyResource` resource instance,
+the `<domain-name>` and `/` are copied verbatim, the `CamelCaseKind` is
+then converted
+using '-' instead of capitalization ('camel-case'), with the first character being assumed to be
+capitalized.  In pseudo code:
+
+```go
+var result string
+for ix := range kindName {
+  if isCapital(kindName[ix]) {
+    result = append(result, '-')
+  }
+  result = append(result, toLowerCase(kindName[ix])
+}
+```
+
+As a concrete example, the resource named `example.com/camel-case-kind` defines resources of Kind `example.com/CamelCaseKind`.
+
+The reason for this is to enable rapid lookup of a `ThirdPartyResource` object given the kind information.
+This is also the reason why `ThirdPartyResource` is not namespaced.
+
+To convert a resource name into the path for the RESTful resource, the name is simply lower-cased.
+`example.com/CamelCaseKind` becomes `example.com/camelcasekind`, see below for more details.
+
+## Usage
+
+When a user creates a new `ThirdPartyResource`, the Kubernetes API Server reacts by creating a new, namespaced
+RESTful resource path.  For now, non-namespaced objects are not supported. As with existing built-in objects
+deleting a namespace, deletes all third party resources in that namespace.
+
+For example, if a user creates:
+
+```yaml
+metadata:
+  name: example.com/cron-tab
+  apiVersion: experimental/v1
+  kind: ThirdPartyResource
+description: "A specification of a Pod to run on a cron style schedule"
+versions:
+  - name: v1
+    apiGroup: stable
+  - name: v2
+    apiGroup: experimental
+```
+
+Then the API server will program in two new RESTful resource paths:
+   * `/thirdparty/example.com/stable/v1/namespaces/<namespace>/crontab/...`
+   * `/thirdparty/example.com/experimental/v2/namespaces/<namespace>/crontab/...`
+
+
+Now that this schema has been created, a user can `POST`:
+
+```json
+{
+   "metadata": {
+     "name": "my-new-cron-object"
+   },
+   "apiVersion": "v1",
```

Comment:
How would kubectl figure out what path to use, say for `kubectl create -f mycron.yaml`?

If sent directly to the API we can automatically populate those fields from the URL.